Gems/Terrain/Code/Source/TerrainSystem/TerrainSystem.cpp

```diff
@@ -177,6 +177,163 @@ bool TerrainSystem::InWorldBounds(float x, float y) const
     return false;
 }
 
+void TerrainSystem::MapPositionToArea(float x, float y,
+    AZStd::unordered_map<AZ::EntityId, TerrainHeightDataLists>& areaIdToPos,
+    AZStd::vector<TerrainHeightDataIterators>& outPositions,
+    size_t outIndex) const
+{
+    AZ::Aabb bounds;
+    AZ::EntityId areaId = FindBestAreaEntityAtPosition(x, y, bounds);
+    auto it = areaIdToPos.find(areaId);
+    if (it == areaIdToPos.end())
+    {
+        auto insertResult = areaIdToPos.insert(AZStd::pair<AZ::EntityId, TerrainHeightDataLists>(areaId, {}));
+        it = insertResult.first;
+        
+        // We maintain iterators to the positions to correspond to the input positions list.
+        // We resize these to the max expected size because we don't want any reallocations happening later.
+        // Reallocations will invalidate our iterators.
+        (*it).second.m_positions.resize(outPositions.size());
+        (*it).second.m_terrainExists.resize(outPositions.size());
```

Comment:
In cases where we have multiple positions split across many area IDs, a bulk query will end up consuming a lot more memory than it needs to.